packages/react/src/hooks/useAuth.ts

```diff
@@ -1,40 +1,43 @@
 import type { ActJWTClaim, GetToken, SignOut } from '@clerk/types';
+import { useCallback, useMemo } from 'react';
 
 import { useAuthContext } from '../contexts/AuthContext';
 import { useIsomorphicClerkContext } from '../contexts/IsomorphicClerkContext';
 import { invalidStateError } from '../errors';
+import IsomorphicClerk from '../isomorphicClerk';
 import { createGetToken, createSignOut } from './utils';
 
-type UseAuthReturn =
-  | {
-      isLoaded: false;
-      isSignedIn: undefined;
-      userId: undefined;
-      sessionId: undefined;
-      actor: null;
-      signOut: SignOut;
-      getToken: GetToken;
-    }
-  | {
-      isLoaded: true;
-      isSignedIn: false;
-      userId: null;
-      sessionId: null;
-      actor: null;
-      signOut: SignOut;
-      getToken: GetToken;
-    }
-  | {
-      isLoaded: true;
-      isSignedIn: true;
-      userId: string;
-      sessionId: string;
-      actor: ActJWTClaim | null;
-      signOut: SignOut;
-      getToken: GetToken;
-    };
+type InitReturn = {
+  isLoaded: false;
+  isSignedIn: undefined;
+  userId: undefined;
+  sessionId: undefined;
+  actor: null;
+  signOut: SignOut;
+  getToken: GetToken;
+};
+
+type LoadedReturn = {
+  isLoaded: true;
+  isSignedIn: false;
+  userId: null;
+  sessionId: null;
+  actor: null;
+  signOut: SignOut;
+  getToken: GetToken;
+};
+
+type UserReturn = {
+  isLoaded: true;
+  isSignedIn: true;
+  userId: string;
+  sessionId: string;
+  actor: ActJWTClaim | null;
+  signOut: SignOut;
+  getToken: GetToken;
+};
 
-type UseAuth = () => UseAuthReturn;
+type UseAuth = () => InitReturn | LoadedReturn | UserReturn;
```

Comment:
Let's avoid creating these three different states and stick to the original type definition to avoid the extra assertions.

Reply:
If I use the original, TS doesn't seem to differentiate them properly and I get this error:

```
Type '() => { isLoaded: false; isSignedIn: any; sessionId: any; userId: any; actor: any; signOut: SignOut; getToken: GetToken; } | { isLoaded: true; isSignedIn: false; sessionId: string; userId: string; actor: any; signOut: SignOut; getToken: GetToken; } | { ...; }' is not assignable to type 'UseAuth'.
      Type '{ isLoaded: false; isSignedIn: any; sessionId: any; userId: any; actor: any; signOut: SignOut; getToken: GetToken; } | { isLoaded: true; isSignedIn: false; sessionId: string; userId: string; actor: any; signOut: SignOut; getToken: GetToken; } | { ...; }' is not assignable to type 'UseAuthReturn'.
        Type '{ isLoaded: true; isSignedIn: false; sessionId: string; userId: string; actor: any; signOut: SignOut; getToken: GetToken; }' is not assignable to type 'UseAuthReturn'.
          Type '{ isLoaded: true; isSignedIn: false; sessionId: string; userId: string; actor: any; signOut: SignOut; getToken: GetToken; }' is not assignable to type '{ isLoaded: true; isSignedIn: true; userId: string; sessionId: string; actor: ActJWTClaim; signOut: SignOut; getToken: GetToken; }'.
            Types of property 'isSignedIn' are incompatible.
              Type 'false' is not assignable to type 'true'.
```